clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/ChainResolver.java

```diff
@@ -72,62 +76,67 @@ public ChainResolver(final OperationsCodec<K, V> codec, Expiry<? super K, ? supe
+              result = null;
+            }
+          } else {
+            Duration duration;
+            try {
+              if(previousResult == null) {
+                duration = expiry.getExpiryForCreation(key, result.getValue());
+                if (duration == null) {
+                  result = null;
+                  continue;
+                }
+              } else {
+                duration = expiry.getExpiryForUpdate(key, new ValueSupplier<V>() {
+                  @Override
+                  public V value() {
+                    return previousResult.getValue();
+                  }
+                }, result.getValue());
+                if (duration == null) {
+                  continue;
                 }
-              }, result.getValue());
-              if (duration == null) {
-                continue;
               }
+            } catch (Exception ex) {
+              LOG.error("Expiry computation caused an exception - Expiry duration will be 0 ", ex);
+              duration = Duration.ZERO;
+            }
+            compacted = true;
+            if(duration.isInfinite()) {
+              expirationTime = Long.MIN_VALUE;
+              continue;
+            }
+            long time = TIME_UNIT.convert(duration.getLength(), duration.getTimeUnit());
+            expirationTime = time + operation.timeStamp();
+            if(now >= expirationTime) {
+              result = null;
             }
-          } catch (Exception ex) {
-            LOG.error("Expiry computation caused an exception - Expiry duration will be 0 ", ex);
-            duration = Duration.ZERO;
-          }
-          compacted = true;
-          if(duration.isInfinite()) {
-            expirationTime = Long.MIN_VALUE;
-            continue;
-          }
-          long time = TIME_UNIT.convert(duration.getLength(), duration.getTimeUnit());
-          expirationTime = time + operation.timeStamp();
-          if(now >= expirationTime) {
-            result = null;
           }
         }
       } else {
         payload.flip();
         chainBuilder = chainBuilder.add(payload);
       }
     }
-    if(result == null) {
-      compacted = keyMatch > 0;
+
+    compacted = (result == null) ? (keyMatch > 0) : (compacted || keyMatch > 1);
+    if(compacted) {
+      if(result != null) {
+        Operation<K, V> resolvedOperation = new PutOperation<K, V>(key, result.getValue(), -expirationTime);
+        ByteBuffer payload = codec.encode(resolvedOperation);
+        chainBuilder = chainBuilder.add(payload);
+      }
+      return new ResolvedChain.Impl<K, V>(chainBuilder.build(), key, result);
     } else {
-      compacted = compacted || keyMatch > 1;
-      Operation<K, V> resolvedOperation = new PutOperation<K, V>(key, result.getValue(), -expirationTime);
-      ByteBuffer payload = codec.encode(resolvedOperation);
-      chainBuilder = chainBuilder.add(payload);
+      return new ResolvedChain.Impl<K, V>(null, key, result);
```

Comment:
I just wanted to avoid rebuilding a non-compacted chain to create the `ResolvedChain.Impl` instance. Passing _null_ was my way of avoiding the rebuilding and the _boolean_ flag indicating whether the `ResolvedChain` has a compacted chain. But with the _boolean_ flag reverted as mentioned in the previous comment I should be able to get rid of this null. Instead of rebuilding the chain I could just pass the origial chain itself here.